text/3557-out-of-tree-test-suite.md

```diff
@@ -0,0 +1,359 @@
+- Feature Name: Out-of-tree test suite
+- Start Date: 2024-01-10
+- RFC PR: [rust-lang/rfcs#3557](https://github.com/rust-lang/rfcs/pull/3557)
+- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)
+
+# Summary
+[summary]: #summary
+
+We propose the creation of an external stand-alone test suite and associated tools for verifying the Rust toolchain that also fits into existing Rust development workflows.
+
+# Motivation
+[motivation]: #motivation
+
+## Testing the artifacts to be distributed
+[testing-the-artifacts-to-be-distributed]: #testing-the-artifacts-to-be-distributed
+
+Industrial use of language toolchains requires assurance that the toolchain that is distributed has been rigorously tested and is guaranteed - even warranted - to work correctly. To achieve this level of assurance, the toolchain is built and packaged for distribution, installed on a completely clean machine, and then tested. This separation of concerns, which mimics the way a user will interact with the toolchain, ensures that the artifacts distributed are the artifacts tested.
+
+Currently, the Rust infrastructure does not support this behavior. Instead, testing validates intermediate states of the Rust toolchain as they exist in the build tree. An intermediate state may contain leftover artifacts from a previous build that could influence the validation. The Rust infrastructure may be directed to produce an malformed or incomplete tool (ex: `rustc` with a missing `libstd`), and yet hold the belief that the tool has successfully passed validation.
```

Comment:
That confidence however relies heavily on the correctness of post-test processing. If there is no post-install verification, then there is no way to detect a post-test bug in the infrastructure, which also includes pilot errors (forgetting to add `rust-std` to a `./x dist` call).

Reply:
The problem of "obvious" post-install errors is easily solved with a few dozen lines of bash running all tools on hello world. For less obvious post-install errors, like a miscompilation, running the normal test suite on the in-tree code (like the current "post-optmization tests" is already done today.
Basically, the thing this OOTTS would be testing that the current test suite isn't testing is `./x dist`, it would be nice to have more tests for it, but writing a test suite for that does not involve moving all UI tests, but instead creating something (a lot smaller) specifically for that.

Reply:
Such a small "install sanity check" test suite can still be in-tree just fine, optionally shipped as a component if you want. That wouldn't require nearly as much work as this RFC and could probably just be a compiler MCP (after talking to the compiler team about it).